dagshub/streaming/dataclasses.py

```diff
@@ -0,0 +1,126 @@
+from dataclasses import dataclass
+from enum import auto, Flag
+from pathlib import Path
+from typing import Optional, Any
+
+try:
+    from functools import cached_property
+except ImportError:
+    from cached_property import cached_property
+
+
+@dataclass
+class StorageAPIEntry:
+    name: str
+    protocol: str
+    list_path: str
+
+    @cached_property
+    def full_path(self):
+        return f"{self.protocol}/{self.name}"
+
+    @cached_property
+    def path_in_mount(self) -> Path:
+        return Path(".dagshub/storage") / self.protocol / self.name
+
+
+@dataclass
+class ContentAPIEntry:
+    path: str
+    # Possible values: dir, file, storage
+    type: str
+    size: int
+    hash: str
+    # Possible values: git, dvc, bucket
+    versioning: str
+    download_url: str
+    content_url: Optional[str]  # TODO: remove Optional once content_url is exposed in API
+
+
+class DagshubPathType(Flag):
+    UNKNOWN = auto()
+    TRACKED_IN_REPO = auto()
+    OUT_OF_REPO = auto()
+    STORAGE_PATH = auto()
+    PASSTHROUGH_PATH = auto()
+
+
+storage_schemas = ["s3:/", "gs:/"]
+
+
+@dataclass
+class DagshubPath:
+    """
+    Class for handling any path used inside the virtual filesystem
+
+    Attributes:
+        fs (DagsHubFilesystem): Filesystem from which this path is assigned
+        absolute_path (Path): Absolute path in the system
+        relative_path (Optional[Path]): Path relative to the root of the encapsulating FileSystem.
+                                        If None, path is outside the FS
+    """
+    # TODO: this couples this class hard to the fs, need to decouple later
+    fs: Any  # Actual type is DagsHubFilesystem, but imports are wonky
+    absolute_path: Optional[Path]
+    relative_path: Optional[Path]
+
+    def __post_init__(self):
+        # Handle storage paths - translate s3:/bla-bla to .dagshub/storage/s3/bla-bla
+        if self.relative_path is not None:
+            str_path = self.relative_path.as_posix()
+            for storage_schema in storage_schemas:
+                if str_path.startswith(storage_schema):
```

Comment:
Won't be neater because I'm handling `s3:/path` and not `s3://path`, so the url schema parsing won't work here and I'll need to add this padding slash in there. Besides, string mangling here is probably faster + I already have tests + docstring, so the thing being done is explained
This is all because the path parameter is a Path and not just a raw string

Reply:
Does that work if I manually try to get access `s3://<bucket>` in my path?